agent/stats/engine.go

```diff
@@ -474,9 +474,8 @@ func (engine *DockerStatsEngine) publishMetrics(includeServiceConnectStats bool)
 	metricsMetadata, taskMetrics, metricsErr := engine.GetInstanceMetrics(includeServiceConnectStats)
 	if metricsErr == nil {
 		metricsMessage := ecstcs.TelemetryMessage{
-			Metadata:                   metricsMetadata,
-			TaskMetrics:                taskMetrics,
-			IncludeServiceConnectStats: includeServiceConnectStats,
```

Comment:
+1

Reply:
Hmm good catch, I think I miss some change, I should have 1) removed this field from the `telemetryMessage` type as well as the condition [here](https://github.com/Realmonia/amazon-ecs-agent/blob/dev/ecs-agent/tcs/client/client.go#L128-L129), or 2) keep use this value. I was planning (but failed) to do 1). I want to do 1) because:

1. ServiceConnect is associated with task. Therefore, having or not having service connect metrics "should" be associated with each "TaskMetric", instead of a "TelemetryMessage", as "TelemetryMessage" contains all tasks on the instance. For Fargate it is a little bit different, since each Fargate instance corresponding to one agent corresponding to one task.
2. EC2 agent collects service connect metrics once per minute, i.e. every 3 telemetryMessages should have 1 message contains service connect metrics. Therefore, when we call `GetInstanceMetrics`, we do need to pass in this parameter to actually **try** to collect service connect metrics. Whether or not a task has service connect metrics is a different story.
3. On the other side of the channel, TCS client, 
    1. when receiving the telemetry message, it calls `publishMetrics` [code](https://github.com/Realmonia/amazon-ecs-agent/blob/dev/ecs-agent/tcs/client/client.go#L130) when either metrics is not disabled, or there is SC metrics, or both. However, when metrics are disabled, [SC metrics are still being collected](https://github.com/Realmonia/amazon-ecs-agent/blob/dev/agent/stats/engine.go#L378-L395), which means in the case that metrics are disabled and there is SC metrics, `getInstanceMetrics` will collect those task metrics of which task has SC enabled, and task metrics will only have SC metrics [here](https://github.com/Realmonia/amazon-ecs-agent/blob/dev/agent/stats/engine.go#L546-L573), and it will form a `telemetryMessage`. The only case, that the condition in channel receiver side skip the publish, is when both metrics is disabled and there is no SC metrics, there will not be a telemetry message at all [here](https://github.com/Realmonia/amazon-ecs-agent/blob/dev/agent/stats/engine.go#L584-L586), so the condition can be removed. In fact, it would not even get to that line - the instance will be an idle instance [here](https://github.com/Realmonia/amazon-ecs-agent/blob/dev/agent/stats/engine.go#L513).
    2. when sending the metrics, it checks whether each `taskMetric` object contains with service connect metric by checking [this](https://github.com/Realmonia/amazon-ecs-agent/blob/dev/ecs-agent/tcs/client/client.go#L221). There is no need for this value. To be more specific, when  `IncludeServiceConnectStats` is set to true, there might be some `taskMetric` in `taskMetrics` contains SC stats; when `IncludeServiceConnectStats` is set to false, there will be no `taskMetric` in `taskMetrics` contains SC stats. As a result, checking if SC stats present in `taskMetric` should be a simpler and direct way to achieve same goal.

By the way, currently we do use this `includeServiceConnectStats` parameter in TCSClient's `metricsToPublishMetricRequests` function under /agent, while Fargate's implementation of TCSClient does not contain it. Again it is looping through each `taskMetric` object and trying to copy its service connect stats, this flag does not provide any value.

Would you prefer 1) removed this field from the `telemetryMessage` type as well as the condition [here](https://github.com/Realmonia/amazon-ecs-agent/blob/dev/ecs-agent/tcs/client/client.go#L128-L129), or 2) keep use this value ?

Reply:
I would prefer 1 as long as we have a clear change in our tests that reflects the code change.

Reply:
Yeah we have getInstance test changes verifying there is no empty telemetry message under all cases and We also have SC integ test to verify cases where metrics are disabled.